whats-that-typeclass-alternative/article.md

```diff
@@ -0,0 +1,414 @@
+deceleration. It isn't required but makes it more easy to understand what
+is going on. To be able to do that, you need to enable the `InstanceSigs` language extension.
+
+Before we can implement `Alternative`, we also need to implement
+`Applicative`:
+
+```haskell
+instance Applicative (Parser i) where
+    pure :: a -> Parser i a
+    pure a = Parser (\is -> (Just a, is))
+    -- with TupleSections could be written as : Parser (Just a,)
+
+    (<*>) :: Parser i (a -> b) -> Parser i a -> Parser i b
+    (Parser f) <*> (Parser p) = Parser $ \is ->
+      let (fM, rest) = f is
+      in case fM of
+           Nothing -> (Nothing, rest)
+           Just f' -> case p rest of
+                        (Just o, rest')  -> (pure $ f' o, rest')
+                        (Nothing, rest') -> (Nothing, rest')
+```
+
+I won't go into the details of `Applicative` here, as there are [excellent other
+articles](https://serokell.io/blog/whats-that-typeclass-applicative) on the idea
+and use of applicative functors. But with that out of the way, we're finally able
+to define the `Alternative` typeclass for our parser.
+
+```haskell
+instance Alternative (Parser i) where
+    empty = Parser (\is -> (Nothing, is))
+    -- with TupleSections could be written as: Parser (Nothing,)
+
+    (<|>) :: Parser i a -> Parser i a -> Parser i a
+    (Parser pa) <|> (Parser pb) = Parser $ \is ->
+      let (aMa, restA) = pa is
+          (aMb, restB) = pb is
+      in case (aMa, aMb) of
+           (Just a, _)       -> (Just a, restA)
+           (Nothing, Just b) -> (Just b, restB)
+           _                 -> (Nothing, is)
+```
+
+But why should we actually want an `Alternative` instance for the parser?
+First of all, because of the possibility to compose multiple `Parser i a` with
+the `<|>` operator.
+
+Another reason is two methods of `Alternative` that we haven't
+discussed so far, which we get for free:
+
+* `some :: f a -> f [a]` which means: _one or more_
+* `many :: f a -> f [a]` which means: _zero of more_
+
+In our case, `some (parseLT 10)` will create a new parser of type
+`parseLT :: Int -> Parser Int [Int]` that expects at least one integer
+that is smaller 10 and then will consume all following numbers which are
+smaller 10 as well.
+
+In case of `many (parseLT 10)`, the type signature looks the same as in the
+`some` case, but it does not fail if the first element in the input isn't a
+number smaller 10. In that case, it will just return the empty list `[]`.
+
+So lets try out these methods:
+
+```none
+λ> evalParser (some $ parseLT 10) [2,3,10,20]
+Right [2,3]
+
+λ> evalParser (many $ parseLT 10) [2,3,10,20]
+Right [2,3]
+
+λ> evalParser (some $ parseLT 10) [10,2,3,20]
+Left "parser failed on [10,2,3,20]"
+
+λ> evalParser (many $ parseLT 10) [10,2,3,20]
+Right []
+```
+
+`some` and `many` are a very common and useful pattern in parser combinators.
+
+If you want to dive deeper into the topic of parsers in Haskell, take a look
```

Comment:
I'd make it evident that it's about parser combinators, as we have other blog posts about parsers too.

BTW, what do you think about changing this blog post to include references to this one too?

Reply:
We can add references from it once we publish the Applicative one and this, yeah. 👍

Reply:
The parser article is linked some lines below, but I like the idea of linking back!